### ISSUE TITLE
proposal buildtest build --config option usage

### ISSUE THREAD

**Issue by shahzebsiddiqui** [buildtesters/buildtest#168]
Current Behavior
-------------------

Usage: ``buildtest build -c /path/to/config.yml`` is path is relative to *toolkit/suite*

The ``--config`` or short option ``-c`` is a choice field auto-populated by buildtest. 

The current behavior few limitation including the following

- Test must be in **toolkit/suite**
- Can only build one test at a time
- There needs to be one sub-directory followed by test so **toolkit/suite/compilers/hello.c.yml**. If yaml file is found more than one directory deep, it causes issue with test generation

Proposal
-----------

The overall objective of this issue to address the current issues by achieving the following

- Capable of building more than one test
- Not be bound to a directory path like **toolkit/suite**
- Change ``--config`` to a string instead of a choice field in argparse 
- Provide means for processing a single configuration or a directory path which may contain multiple configuration files.

When user downloads the test such as the tutorial example by running ``buildtest get https://github.com/HPC-buildtest/tutorials.git`` the test are stored in 

Config usage
----------------------

- Build by test name: ``buildtest build -c hello.yml``

- Absolute Path: ``buildtest build -c ~/.buildtest/site/github.com/HPC-buildtest/tutorials/compilers/hello.f.yml``

- Relative Path to site directory: ``buildtest build -c github.com/HPC-buildtest/tutorials/compilers/hello.f.yml``

- Build by directory path (abspath): ``buildtest build -c ~/.buildtest/site/github.com/HPC-buildtest/tutorials/compilers`` 

- Build by directory path (relative path): ``buildtest build -c github.com/HPC-buildtest/tutorials/compilers``

For scope of PR the above features should be addressed to be significant change from current implementation. It is sufficient to resolve the test configuration correctly and simply print out the full path to YAML file and not actually parse, build or run any test. This can be fixed in subsequent PRs when working on the build component.

Additional options to CLI usage will be addressed in another issue.

**Comment by vsoch** [comment]
I like the ability for buildtest to discover recipes in a directory, but it should also be possible to do

```bash
buildtest -c /path/to/recipe.yml
```
and just have it work. As a comparison, image if Singularity or docker required the user to register a path to Dockerfiles or Singularity recipes before running `docker build` or `singularity build`. It would be hugely not intuitive and frustrating. What seems intuitive is to accept a filename OR a directory, and then allow filtering based on a name (e.g., internally using something fnmatch). So these would be valid:

```bash
buildtest build -c /path/to/file.yml   # explicitly build file.yml
buildtest build -d /path/to/folder  # build all recipes discovered in folder
buildtest build -d /path/to/folder --pattern file*.yml  # build all recipes in folder that start with file*
```
In the case that a path isn't provided, e.g., let's say the user does:

```bash
buildtest build site/tutorials
```
Then we would then check for the path in the build.testdir. If it's not there, then there is an error. Are there any missed cases for these mentioned above? I don't think we should restrict directories to test root, that would make life really hard because you can't run a file / folder off the bat.

But test root is useful! If we keep the TEST_ROOT list, if the user just issues:

```bash
buildtest build
```
And it would run all tests for folders defined in TEST_ROOT. If the user wanted to do the same but match a pattern, that would work too:

```bash
# assumes paths in TEST_ROOT that match pattern mpi*.yml
buildtest build --pattern "mpi*.yml"
```
I think before implementing this change it would make sense to rename toolkit/suite to be just site, I can open a PR to do a first shot at this. I'm going to try to work in small PRs so they are easier to review and then merge into devel - large ones are dangerous!

**Comment by shahzebsiddiqui** [comment]
@vsoch if it's ok I will assign this to you. I know this will require some code refactoring and for scope of this issue let's just focus on the command usage. The YAML schema change will follow after this issue that's covered in #135

**Comment by vsoch** [comment]
The "site" folder has been already integrated, and the issue to support:

```
buildtest -c /path/to/recipe.yml
```
will be addressed by #191.   We will need to (after this is merged, and likely more work done) discuss again providing a folder and allowing for recipe discoverability.

**Comment by shahzebsiddiqui** [comment]
Yes you are correct @vsoch. I think once your PR #191 is merged we can close this issue and work on the build component and YAML schema. Later we can discuss additional features to configuration option described in #193

**Comment by vsoch** [comment]
This was finished as #191 was merged, closing.

**Comment by shahzebsiddiqui** [comment]
Thanks for closing ticket 👍
